sos/report/plugins/roce.py

```diff
@@ -0,0 +1,87 @@
+# Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+# This file is part of the sos project: https://github.com/sosreport/sos
+#
+# This copyrighted material is made available to anyone wishing to use,
+# modify, copy, or redistribute it subject to the terms and conditions of
+# version 2 of the GNU General Public License.
+#
+# See the LICENSE file in the source distribution for further information.
+
+from sos.report.plugins import Plugin, IndependentPlugin
+
+
+class RoCE(Plugin, IndependentPlugin):
+
+    short_desc = 'RoCE (RDMA over Converged Ethernet) information'
+
+    plugin_name = 'roce'
+    profiles = ('hardware',)
+    # rdma installed with iproute2
+    packages = ('iproute2', 'infiniband-diags')
+
+    def setup(self):
+
+        IB_SYS_DIR = "/sys/class/infiniband/"
+        ibs = self.listdir(IB_SYS_DIR) if self.path_isdir(IB_SYS_DIR) else []
+        for ib in ibs:
+
+            # cma_roce_mode and cma_roce_tos will not work for bond devices
+            # ibstat is used to check if the device exist or not, form bond
+            # the command will return an erro:
+            # stat of IB device <device> failed: No such file or directory
+            skip_cma_roce = False
+            ibstat_out = self.exec_cmd(f'ibstat -s {ib}')
+            if ibstat_out['status'] != 0:
+                skip_cma_roce = True
+
+            # dump ECN configuration
+
+            for port in self.listdir(IB_SYS_DIR + ib + "/ports"):
+                # skip Infiniband and IWARP devices
+                try:
+                    p = open(IB_SYS_DIR + ib + "/ports/" + port +
+                             "/link_layer")
+                except IOError:
+                    continue
+                link_layer = p.readline()
+                p.close()
+                if link_layer != "Ethernet\n":
+                    continue
+
+                # dump counters
+                roce_counters = IB_SYS_DIR + ib + "/ports/" + port + \
+                    "/counters"
+                self.add_copy_spec([roce_counters])
+
+                # dump HW counters
+                roce_hw_counters = IB_SYS_DIR + ib + "/ports/" + port + \
+                    "/hw_counters"
+                self.add_copy_spec([roce_hw_counters])
+
+                # dump gids
+                gids = IB_SYS_DIR + ib + "/ports/" + port + "/gids"
+                self.add_copy_spec([gids])
+
+                # dump gid attributes
+                gid_attrs = IB_SYS_DIR + ib + "/ports/" + port + "/gid_attrs"
+                self.add_copy_spec([gid_attrs])
+
+                # dump ECN configuration
+
+                netsys = IB_SYS_DIR + ib + "/device/net"
+                for netdev in self.listdir(netsys):
```

Comment:
**Problem**: The variable `netsys` is not checked if it is a directory before listing its contents. This could lead to an error if `netsys` is not a directory.

**Suggested Change**: Check if `netsys` is a directory before listing its contents.


**Severity (1 - 4)**: 2 - MINOR

**Lines**: 71-72

```suggestion
                netsys = IB_SYS_DIR + ib + "/device/net"
                if self.path_isdir(netsys):
                    for netdev in self.listdir(netsys):
                        # ...
```

-----

<div align="right">
    <small><i>Generated by <b>Llama 3.1 405b</b></i></small>
</div>
<br />


<div align="right">
  <small>Was this helpful?</small>
  <a href="https://codecritic-feedback.nvidia.com?vote=1&comment_guid=1fded783-6eea-498b-afd9-f4ad367c9d82&is_github=true" target="_blank">
    <button>👍</button>
  </a>
  <a href="https://codecritic-feedback.nvidia.com?vote=-1&comment_guid=1fded783-6eea-498b-afd9-f4ad367c9d82&is_github=true" target="_blank">
    <button>👎</button>
  </a>
</div>